clang/lib/Parse/Parser.cpp

```diff
@@ -1100,30 +1100,32 @@ Parser::DeclGroupPtrTy Parser::ParseDeclOrFunctionDefInternal(
   // C99 6.7.2.3p6: Handle "struct-or-union identifier;", "enum { X };"
   // declaration-specifiers init-declarator-list[opt] ';'
   if (Tok.is(tok::semi)) {
-    auto LengthOfTSTToken = [](DeclSpec::TST TKind) {
-      assert(DeclSpec::isDeclRep(TKind));
-      switch(TKind) {
-      case DeclSpec::TST_class:
-        return 5;
-      case DeclSpec::TST_struct:
-        return 6;
-      case DeclSpec::TST_union:
-        return 5;
-      case DeclSpec::TST_enum:
-        return 4;
-      case DeclSpec::TST_interface:
-        return 9;
-      default:
-        llvm_unreachable("we only expect to get the length of the class/struct/union/enum");
+    auto GetAdjustedAttrsLoc = [&]() {
```

Comment:
Lambda is a little weird here?  Rather than just setting a variable?

Something like: 
```
SourceLocation CorrectLocationForAttributes{};

if (auto TKind = DS.getTypeSpecType(); DeclSpec::isDeclRep(TKind)) {
...
```

Reply:
the location + if has a little less mental load when reading this?  And I'd hope with the below suggestions this stays pretty small/gets smaller.